python/lsst/ts/mtaos/mtaos.py

```diff
@@ -762,6 +827,118 @@ async def do_resetOffsetDOF(self, data: salobj.type_hints.BaseDdsDataType) -> No
             await self.pubEvent_degreeOfFreedom()
             await self.pubEvent_mirrorStresses()
 
+    async def do_stopClosedLoop(self):
+        """Stop the closed loop operation."""
+        self.assert_enabled()
+        self._logExecFunc()
+
+        if self.close_loop_task.done():
+            self.log.warning("Closed loop already stopped.")
+            return
+
+        self.closed_loop_task.cancel()
+        await self.evt_closedLoopState.set_write(state=ClosedLoopState.IDLE)
+        try:
+            await self.closed_loop_task
+        except asyncio.CancelledError:
+            pass
+
+    def do_startClosedLoop(self):
+        """Start the closed loop operation."""
+        self.assert_enabled()
+        if self.closed_loop_task.done():
+            self.closed_loop_task = asyncio.create_task(self.run_closed_loop())
+        else:
+            self.log.info("Closed loop already running, nothing to do.")
+
+    async def run_closed_loop(self):
+        """Closed loop operation."""
+        self._logExecFunc()
+
+        prev_elevation = None
+        async with salobj.Remote(
+            "MTOODS" if self.camera_name == "LSSTCam" else "CCOODS"
+        ) as oods:
+            oods.evt_imageInOODS.flush()
+            while self.summary_state == salobj.State.ENABLED:
+                try:
+                    await self.evt_closedLoopState.set_write(
+                        state=ClosedLoopState.WAITING_IMAGE
+                    )
+
+                    image_in_oods = await oods.evt_imageInOODS.next(flush=False)
+                    self.log.info(
+                        f"Image {image_in_oods.obsid} {image_in_oods.raft} {image_in_oods.sensor} ingested."
+                    )
+
+                    await self.evt_closedLoopState.set_write(
+                        state=ClosedLoopState.PROCESSING
+                    )
+
+                    self._execute_wavefront_estimation(
+                        data=salobj.cmd_runWEP.DataType(
```

Comment:
So, `salobj` does not have `cmd_runWEP`, you probably meant `self` here. However, see my comment above about the `_execute_wavefront_extimation` method having explicit arguments instead of receiving an obfuscated `data` object.